easy_transformer/hook_points.py

```diff
@@ -1,149 +1,216 @@
-# Import stuff
 import logging
-from typing import Callable, Union, Optional, Sequence
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Callable, Optional, Sequence, Union
+
 import torch
 import torch.nn as nn
 
-from functools import *
 
-# %%
-# Define type aliases
-NamesFilter = Optional[Union[Callable[[str], bool], Sequence[str]]]
+class NamesFilter(ABC):
+    @abstractmethod
+    def is_included(self, name: str) -> bool:
+        pass
+
+
+class FunctionalNameFilter(NamesFilter):
+    def __init__(self, f: Callable[[str], bool]):
+        self.f = f
+
+    def is_included(self, name: str) -> bool:
+        return self.f(name)
+
+
+class ConstantNameFilter(NamesFilter):
+    def __init__(self, names: Sequence[str]):
+        self.names = set(names)
+
+    def is_included(self, name: str) -> bool:
+        return name in self.names
+
+
+# TODO did i fuck up by not mkaing this optional?
 
 # %%
-# A helper class to get access to intermediate activations (inspired by Garcon)
-# It's a dummy module that is the identity function by default
-# I can wrap any intermediate activation in a HookPoint and get a convenient
-# way to add PyTorch hooks
+
+
+# TODO make sure this is used in all the right places
+class HookDirection(Enum):
```

Comment:
I think this is way uglier? Or at least, people will now do add_hook(dir=1) which seems way worse than dir="fwd"? (Unless you expect them to import + use this enum)

Reply:
bcbb687